### PR TITLE
Header Search

### PR DESCRIPTION
**Pull request by MadameSheema**
Solves: https://github.com/franksparks/BookJourney/issues/10

Screenshot 2024-07-26 at 20 16 41

## Summary

In this PR we are implementing the header search box, currently:

- Performs a search each time that you input a letter
- Displays the first 5 results
- Displays `See all results` at the end

What it has not been developed yet:
- No action has been developed yet for when a user clicks in one of the options, so currently, clicking an option it will perform an extra search with the selected book, this works like this since is the default behaviour of the used component.

## Implementation

To implement the search bar, it has been used the [MUI autocomplete](https://mui.com/material-ui/react-autocomplete/) with some modifications to make it behave and look as we expect.

- The built-int filtering has been disabled
- The popup icon has been removed
- We use the input change to perform a query to the googlebooks api
- We just open the dropdown with the options when the user has at least entered one character and there are books available in the response
- We have disabled all the animations

In this implementation we are using two different states:
- options, where we'll store the results of the api call to be displayed
- inputValue, the string/characters typed in the input (the user query)

Each time there is a change in the input, we use the entered query to execute a query to googleBooks with the enterered value and we modify it to comply with the format the component is expecting to be able to display it. 

## To improve

As currently we are querying the API as we write, there is a bit ugliness in the way we are displaying the results, because even when you stop typing you can see how these might change. 

In order to improve it, once we have the core components ready, it will be nice to follow-up this and use debounce to fix it.

### CODE REVIEW

File: components/header.tsx [hunk 1]
```diff
@@ -24,6 +25,7 @@ export const Header = () => {
           <Loader className="h-6 w-6 text-white animate-spin" />
         </ClerkLoading>
         <ClerkLoaded>
+        <PreviewSearchBox/>
```
Comment (franksparks) [inline components/header.tsx:28]: As _PreviewSearchBox_ component is placed after the loading of _Clerk_, during a screen refresh the search box disappears. 

Component should be placed above _ClerkLoading_ tag.

**Comment by franksparks** [comment]
With 100% zoom, "See all results" link is not visible.

Screenshot 2024-07-27 at 12 17 56

**Comment by franksparks** [comment]
Should "Powered by Google" message be displayed?

**Comment by franksparks** [comment]
Successive searches mixes the results
- Search 1 - "Ubik" - 5 results displayed
- Search 2 - "Liu" - 8 results displayed (3 of the first search and 5 from the current one)

**Comment by MadameSheema** [comment]
> With 100% zoom, "See all results" link is not visible.

I see that the scroll bar appears in the screenshot, did you scrolled to the bottom of the results and the `See all results` was not present?

> Should "Powered by Google" message be displayed?

I forget to add that point on the PR, that is something I want to talk with the team since with the current design, the powered by google is not looking good/

> Successive searches mixes the results
> Search 1 - "Ubik" - 5 results displayed
> Search 2 - "Liu" - 8 results displayed (3 of the first search and 5 from the current one)

I would love to test that with you, I entered different values during the development and I didn't find anything weird.

>  This error appeared on Console:

That error is the ones that appears when you delete all the the entered query, this is because as the way the component works, each input is used as query, as when deleting is an empty one, there is no result resulting on that mapping error. 

I'll try to fix it :)

**Review by franksparks — APPROVED** [review]
LGTM

File: lib/volumes.ts [hunk 2]
```diff
@@ -17,9 +17,9 @@ export type VolumesResponse = {
   totalItems: number
 };
 
-export const searchVolumes = async (search: string, index: number) => {
+export const searchVolumes = async (search: string, index: number, maxResults: number) => {
   const books = await fetch(
-    `https://www.googleapis.com/books/v1/volumes?q=${search}&startIndex=${index.toString()}`
+    `https://www.googleapis.com/books/v1/volumes?q=${search}&maxResults=${maxResults}&startIndex=${index.toString()}`
```
Comment (vedderzeznick) [inline lib/volumes.ts:22]: this google URL should be part of a "Constants" file where we have all this kind of info, same thing should happen with the google API token

**Pull request merged**